python/cudf/cudf/core/udf/groupby_typing.py

```diff
@@ -0,0 +1,226 @@
+SUPPORTED_GROUPBY_NUMBA_TYPES = [types.int64, types.float64]
+SUPPORTED_GROUPBY_NUMPY_TYPES = [
+    numpy_support.as_dtype(dt) for dt in SUPPORTED_GROUPBY_NUMBA_TYPES
+]
+
+
+class Group(object):
+    def __init__(self, group_data, size, index, dtype, index_dtype):
+        self.group_data = group_data
+        self.size = size
+        self.index = index
+        self.dtype = dtype
+        self.index_dtype = index_dtype
+
+
+class GroupType(numba.types.Type):
+    def __init__(self, group_scalar_type, index_type=index_default_type):
+        self.group_scalar_type = group_scalar_type
+        self.index_type = index_type
+        self.group_data_type = types.CPointer(group_scalar_type)
+        self.size_type = types.int64
+        self.group_index_type = types.CPointer(index_type)
+        super().__init__(
+            name=f"Group({self.group_scalar_type}, {self.index_type})"
+        )
+
+
+@typeof_impl.register(Group)
+def typeof_group(val, c):
+    return GroupType(
+        numba.np.numpy_support.from_dtype(val.dtype),
+        numba.np.numpy_support.from_dtype(val.index_dtype),
+    )  # Identifies instances of the Group class as GroupType
+
+
+# The typing of the python "function" Group.__init__
+# as it appears in python code
+@type_callable(Group)
+def type_group(context):
+    def typer(group_data, size, index):
+        if (
+            isinstance(group_data, types.Array)
+            and isinstance(size, types.Integer)
+            and isinstance(index, types.Array)
+        ):
+            return GroupType(group_data.dtype, index.dtype)
+
+    return typer
+
+
+@register_model(GroupType)
+class GroupModel(models.StructModel):
+    def __init__(
+        self, dmm, fe_type
+    ):  # fe_type is fully instantiated group type
+        members = [
+            ("group_data", types.CPointer(fe_type.group_scalar_type)),
+            ("size", types.int64),
+            ("index", types.CPointer(fe_type.index_type)),
+        ]
+        models.StructModel.__init__(self, dmm, fe_type, members)
+
+
+call_cuda_functions: Dict[Any, Any] = {}
+
+
+def _register_cuda_reduction_caller(func, inputty, retty):
+    cuda_func = cuda.declare_device(
+        f"Block{func}_{inputty}", retty(types.CPointer(inputty), types.int64)
+    )
+
+    def caller(data, size):
+        return cuda_func(data, size)
+
+    if call_cuda_functions.get(func.lower()) is None:
+        call_cuda_functions[func.lower()] = {}
+
+    type_key = (retty, inputty)
+    call_cuda_functions[func.lower()][type_key] = caller
+
+
+def _register_cuda_idxreduction_caller(func, inputty):
+    cuda_func = cuda.declare_device(
+        f"Block{func}_{inputty}",
+        types.int64(
+            types.CPointer(inputty), types.CPointer(types.int64), types.int64
+        ),
+    )
+
+    def caller(data, index, size):
+        return cuda_func(data, index, size)
+
+    # only support default index type right now
+    type_key = (index_default_type, inputty)
```

Comment:
Is there a way we can fail loudly if the index type is not the default?

Reply:
Added a check inside the lowering at https://github.com/rapidsai/cudf/pull/11452/commits/33109f5a886b10b9164c831f3f0e55578cafed1c . IIUC this method of doing things should make it so that the error is only thrown if numba is trying to compile a UDF containing `idxmax` or `idxmin`.